[zannloo] iP

zannloo opened the pull request:
# TehO

> "TehO is hot tea (without milk) with sugar." - [HoneyKids Editorial](https://honeykidsasia.com/author/hkeditorial/)

TehO helps you remember things you need to do. It is, 
- *text-based*
- **easy** to use
- ~~FAST~~ SUPER FAST to learn

All you need to do is, 
1. download it from here.
2. double-click it. 
3. add your tasks.
4. let TehO manage your tasks for you.

Features:
- [ ] Managing tasks
- [ ] Managing deadline
- [ ] Managing events
- [x] and a **really cute** chatbot for you :heart_eyes::kissing_heart::smiling_face_with_three_hearts:

If you Java programmer, you can use it to practice Java too. Here's the ```main``` method:
```ruby
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
add TehO.java

Commit:
Improved TehO by allowing it echo commands entered by user and exit with command bye.

Commit:
added methods addTask and listTask to TehO.java

Commit:
added new methods to mark and unmark tasks in TehO.java

Commit:
add 3 new class that inherit from Task class

Commit:
add Task.java (for lvl3) and separated toString() method and getStatusIcon() method

Commit:
included todo, deadline and event (incompleted)

Commit:
add support for tracking three types of tasks

Commit:
Use Exceptions to handle two types of errors shown in the example (Minimal).

Commit:
Add support for deleting tasks from the list and switching to ArrayList. (draft1)

Commit:
reverting back to Lvl4 because I missed out A-TextUiTesting

Commit:
editing the runtest.sh for test using the I/O redirection technique

Commit:
Use Expressions to handle two types of errors shown in the website.

Commit:
Add support for deleting tasks from the list and switching to ArrayList (draft1).

Commit:
added the close scanner method (draft 2 of lvl6)

Commit:
lvl 6 draft 3

Commit:
Add support for deleting tasks from the list and corrected old mistakes of not showing the time.

Commit:
last edits for Level 6.

Commit:
final changes and adding the @Override

Commit:
Add teho.txt

TehO class: Update to enable saving of tasks

Commit:
Unsure if the previous reverted commit was undone

Commit:
Store and accept date in a new format

Commit:
Update to have more OOP

Add Ui class: deals with interactions with the user
Add Storage class: deals with loading tasks from the file and saving tasks in the file
Add Parser class: deals with making sense of the user command
Add TaskList class: contains the task list e.g., it has operations to add/delete tasks in the list

Commit:
Organize classes into 2 packages, teho.exceptions and teho.main

Commit:
Merge commit 'ed6d4d25bd4e34461919ffe785a6f10f4c087552'

* commit 'ed6d4d25bd4e34461919ffe785a6f10f4c087552':
  Bump gradle and lib version
  Add Gradle support

Commit:
Add ToDoTest and DeadlineTest to test behaviour of code using JUnit

Commit:
Change build.gradle for A-Jar increment

Commit:
Add JavaDoc

Commit:
Add new feature to allow users find a task by searching for a keyword

Commit:
Tweak the code to comply with a coding standard

Commit:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Tweak the code to comply with a coding standard

Commit:
commit Storage class

Commit:
Merge branch 'branch-Level-9'

* branch-Level-9:
  commit Storage class

Commit:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc

# Conflicts:
#	src/main/java/teho/main/Storage.java
#	src/main/java/teho/main/TehO.java
#	text-ui-test/data/teho.txt

Commit:
Fix some errors in Parser class

Commit:
Edits for level 10

alvinlim277 commented on src/main/java/teho/main/TehO.java:
```diff
@@ -0,0 +1,199 @@
+     */
+    public String addToDo(String userCommand) {
+        try {
+            if (userCommand.length() < 5) {
```
Not sure I like the use of a magic number here even though it's explained in the comments.
Same applies to all the methods that parse user input in this class.

alvinlim277 commented on src/main/java/teho/main/Task.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * Represents a task with only a description and status.
+ */
+public abstract class Task {
```
I like how Task is abstract so it can't be accidentally instantiated.

alvinlim277 commented on src/main/java/teho/main/TehO.java:
```diff
@@ -0,0 +1,199 @@
+                returnMessage = ui.generateList(taskList);
+            } else if (userCommand.startsWith("mark")) {
+                markTask(userCommand);
+                returnMessage = ui.generateUnmarkTaskMessage(taskList.getTask(parseInt(userCommand.split(" ")[1]) - 1));;
```
Not sure I like how complicated the returnMessage expressions are, especially with multiple layers of nested parenthesis.  It might be better for readability if it was done in steps.
```suggestion
int index = parseInt(userCommand.split(" ")[1]) - 1));
Task taskUnmarked = taskList.getTask(index);
returnMessage = ui.generateUnmarkTaskMessage(taskUnmarked);
```
(Something like this maybe? Assuming I managed to understand the code correctly...)

alvinlim277 commented on src/main/java/teho/main/TehO.java:
```diff
@@ -0,0 +1,199 @@
+            if (userCommand.length() < 9) {
+                throw new EmptyDeadlineDescriptionException();
+            }
+            String commandWithDate = userCommand.substring(9); //"deadline " 9 index
+            String cDeadline = commandWithDate.split(" /by ")[0]; //just description
+            LocalDate byDate = LocalDate.parse(commandWithDate.split(" /by ")[1]); //just byDate
+            Task task = new Deadline(cDeadline, byDate);
+            this.taskList.add(task);
```
Really like the splitting of the Task generation into steps here.

alvinlim277 commented on src/main/java/teho/main/TehO.java:
```diff
@@ -0,0 +1,199 @@
+            if (userCommand.length() < 6) {
+                throw new EmptyEventDescriptionException();
+            }
+            String commandWithDate = userCommand.substring(6); //"todo " 6 index
+            String cEvent = commandWithDate.split(" /from ")[0]; //just description
+            String dates = commandWithDate.split(" /from ")[1]; //bothDates
+            LocalDate fromDate = LocalDate.parse(dates.split(" /to ")[0]); //just fromDate
+            LocalDate toDate = LocalDate.parse(dates.split(" /to ")[1]); //just toDate
+            Task task = new Event(cEvent, fromDate, toDate);
+            this.taskList.add(task);
```
Really like the splitting of the Task generation into steps here, especially when the Event takes so many pieces of data to create (applies to the above methods as well).

alvinlim277 commented on src/main/java/teho/main/TehO.java:
```diff
@@ -0,0 +1,199 @@
+            }
+            String commandWithDate = userCommand.substring(6); //"todo " 6 index
+            String cEvent = commandWithDate.split(" /from ")[0]; //just description
+            String dates = commandWithDate.split(" /from ")[1]; //bothDates
+            LocalDate fromDate = LocalDate.parse(dates.split(" /to ")[0]); //just fromDate
+            LocalDate toDate = LocalDate.parse(dates.split(" /to ")[1]); //just toDate
```
I like how clear the variable naming is (so much so that the comments are a bit unecessary).

alvinlim277 commented on src/main/java/teho/main/TehO.java:
```diff
@@ -0,0 +1,199 @@
+                throw new EmptyDeadlineDescriptionException();
+            }
+            String commandWithDate = userCommand.substring(9); //"deadline " 9 index
+            String cDeadline = commandWithDate.split(" /by ")[0]; //just description
```
'cDeadline' could be a bit of a misleading name, assuming it refers to the description of the task.

alvinlim277 commented on src/main/java/teho/main/Ui.java:
```diff
@@ -0,0 +1,145 @@
+     *
+     * @return String representation of hello message
+     */
+    public static String generateHelloMessage() {
```
I like the naming of the methods in this class, they are consistent and clear to the reader.

alvinlim277 reviewed (commented):
LGTM! Just a few nits here and there regarding code quality. The code is really well documented (both JavaDocs and comment wise).

Commit:
Edits

Saezenn commented on src/main/java/teho/main/Event.java:
```diff
@@ -0,0 +1,52 @@
+     *
+     * @param description Description of task with a specific time frame.
+     * @param fromDate Beginning date of the task with a specific time frame.
+     *  @param toDate End date of the task with a specific time frame.
```
Overall, the javadocs are well organised, but there seems to be an extra spacing here.

Saezenn commented on src/main/java/teho/main/Event.java:
```diff
@@ -0,0 +1,52 @@
+     * @param fromDate Beginning date of the task with a specific time frame.
+     *  @param toDate End date of the task with a specific time frame.
+     */
+    public Event(String description, LocalDate fromDate, LocalDate toDate) {
```
In the Event class constructor's Javadoc comment, you can consider specifying that the fromDate should be before or equal to the toDate. This would clarify the expected input conditions.

Saezenn commented on src/main/java/teho/main/Event.java:
```diff
@@ -0,0 +1,52 @@
+    public Event(String description, LocalDate fromDate, LocalDate toDate) {
+        super(description);
+        this.fromDate = fromDate;
+        this.toDate = toDate;
```
In the constructor, it may be a good practice to check if fromDate is after toDate and throw an exception if it is, to prevent creating an event with an invalid date range.

Saezenn commented on src/main/java/teho/main/Parser.java:
```diff
@@ -0,0 +1,44 @@
+package teho.main;
```
Can organize the import statements in alphabetical order. This convention makes it easier to locate specific classes and packages.

Saezenn commented on src/main/java/teho/main/Parser.java:
```diff
@@ -0,0 +1,44 @@
+        if (sections[0].equals("T")) {
+            taskToAddFirst = new ToDo(sections[2]);
+            taskList.add(taskToAddFirst);
+            if (sections[1].equals("1")) {
```
You can simplify your code by creating the task first and then checking if it should be marked as done. This eliminates code duplication and makes it easier to read. (i.e repeating the if sections[1] == 1 check)

Saezenn reviewed (commented):
Looks good! 

Some minor adjustments can be made, but otherwise very organised and clear code!

Commit:
Generate a new JAR file using Gradle

Commit:
Update assertions

Include assertions in Storage constructor, saveTask method in the Storage class to verify that filePath and taskList are not null.
Include assertion in remove method in the TaskList class to verify that taskNumber is not negative.
Include assertions in markTask, unmarkTask and delete methods in TehO class to verify that taskNumber is not negative .

Commit:
Improve code quality

Having fixed sizes for todo, deadline, event and find to avoid magic numbers which improves readability.
Simplifying returnMessage to improves readability.

zannloo committed:
Merge pull request #2 from zannloo/branch-A-Assertions

Update assertions

Commit:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Update assertions

# Conflicts:
#	src/main/java/teho/main/TehO.java

zannloo committed:
Merge pull request #1 from zannloo/branch-A-CodeQuality

Improve code quality

Commit:
Edits

Commit:
Add sort tasks alphabetically feature

Currently, there is no feature to sort tasks in alphabetical order.

Sorting tasks according to alphabetical order can enhance visibility for the user.

It is done through a method sortListAlphabetically so that it is in the same command format as the other commands.

It is done this way so that users can easily learn how to use the command.

Commit:
Edits

Change some phrases used by TehO chatbot for the personality touch.

Commit:
Gives credit for reused work

Commit:
Updates for UserGuide

Commit:
Edits for the UserGuide

Commit:
Removes some comments and includes edits for a feature

Removes some comments for clarity.

Edits to the feature "list" and "sortAlphabetically" such that
the chatbot will return a message when there is no tasks in list.

Commit:
Edits for clarity

Commit:
Edits
